viewer/udSample.h

```diff
@@ -18,27 +24,26 @@ struct udSampleRenderInfo
   struct SDL_Texture *pSDLTexture;
 };
 
-typedef void (udSampleInit)(void **ppSampleData, const udSampleRenderInfo &info);
-typedef void (udSampleDeinit)(void *pSampleData);
-typedef void (udSampleRender)(void *pSampleData, const udSampleRenderInfo &info);
 
-struct udSample
+// Base class for samples
+class udSample
 {
-  const char *pName;
-
-  udSampleInit *pInit; // Called when the Sample is 'activated'
-  udSampleDeinit *pDeinit; // Called when the Sample is 'deactivated'
-  udSampleRender *pRender; // Called each frame
+public:
+  udSample() { samples.push_back(this); }
+  virtual ~udSample() {}
+
+  virtual const char *GetName() const = 0;                          // Return the name of the sample
+  virtual udError Init(udSampleRenderInfo &info) = 0;               // Initialise for running the sample
+  virtual udError Deinit() = 0;                                     // Tear-down resources to run another sample, or to re-run from the start
```

Comment:
The return type is unnecessary, `main.cpp` doesn't use either of these. `udError` is probably more than is required, a simple `bool` would probably suffice.

Reply:
Yeah it was just for future robustness but possibly true, errors on Deinit's are generally hard to make use of